File: opspilot-enterprise/packages/shared-types/src/interaction.ts
import type { IntentDomain, RiskLevel } from "./intent";

export type InteractionKind = "clarify" | "approve";
export type ApprovalScope = "once" | "session";
export type ApprovalDecisionOutcome = "approved" | "rejected" | "expired";

export interface ResourceRef {
  type: string;
  id: string;
  name: string;
}

export interface ResourceScope {
  environment: string;
  resources: ResourceRef[];
}

export interface ClarifyRecord {
  interaction_id: string;
  run_id: string;
  question: string;
  choices: string[];
  allow_free_text: boolean;
  reason_code: "ambiguous_intent" | "missing_slot" | "conflicting_resource" | "unsafe_default";
  status: "pending" | "answered" | "approved" | "rejected" | "expired";
  expires_at: string;
  created_at: string;
  created_by: string;
  selected_choice?: string | null;
  free_text?: string | null;
  responded_at?: string | null;
  responded_by?: string | null;
}

export interface ApprovalRecord {
  approval_id: string;
  run_id: string;
  summary: string;
  domain: IntentDomain;
  action: string;
  risk_level: RiskLevel;
  resource_scope: ResourceScope;
  command_preview: string[];
  plan_steps: string[];
  rollback_plan: string[];
  allowed_scopes: ApprovalScope[];
  status: "pending" | "answered" | "approved" | "rejected" | "expired";
  decision?: ApprovalDecisionOutcome | null;
  final_scope?: ApprovalScope | null;
  comment?: string | null;
  approved_by?: string | null;
  approved_at?: string | null;
  expires_at: string;
  created_at: string;
  created_by: string;
}

export interface ResumeCardData {
  checkpoint_id: string;
  run_id: string;
  last_safe_step?: number | null;
  resume_from?: string | null;
  idempotency_key: string;
  rollback_available: boolean;
}

export interface AuditTimelineData {
  run_id: string;
  operator?: string | null;
  decision_chain: string[];
  tool_outputs: string[];
  events: Array<{
    event_id: string;
    event_type: string;
    summary: string;
    created_at: string;
    actor_type: string;
    actor_id: string;
    step_no: number;
  }>;
}

File: opspilot-enterprise/packages/shared-schema/src/opspilot_schema/interaction.py
from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field

from .intent import IntentDomain, RiskLevel

InteractionKind = Literal["clarify", "approve"]
ClarifyReasonCode = Literal[
    "ambiguous_intent",
    "missing_slot",
    "conflicting_resource",
    "unsafe_default",
]
InteractionStatus = Literal["pending", "answered", "approved", "rejected", "expired"]
ApprovalScope = Literal["once", "session"]
ApprovalDecisionOutcome = Literal["approved", "rejected", "expired"]


class ClarifyChoice(BaseModel):
    id: str
    label: str


class ClarifyRequest(BaseModel):
    interaction_id: str
    run_id: str
    question: str
    choices: list[str] = Field(default_factory=list, max_length=4)
    allow_free_text: bool = True
    reason_code: ClarifyReasonCode = "ambiguous_intent"
    expires_at: str
    created_at: str
    created_by: str = "system"


class ClarifyResponse(BaseModel):
    interaction_id: str
    selected_choice: Optional[str] = None
    free_text: Optional[str] = None
    submitted_at: str
    responded_by: str = "user"


class ClarifyCreateRequest(BaseModel):
    run_id: str
    question: str
    choices: list[str] = Field(default_factory=list, max_length=4)
    allow_free_text: bool = True
    reason_code: ClarifyReasonCode = "ambiguous_intent"
    expires_in_seconds: int = 900


class ClarifyRecord(ClarifyRequest):
    status: InteractionStatus = "pending"
    selected_choice: Optional[str] = None
    free_text: Optional[str] = None
    responded_at: Optional[str] = None
    responded_by: Optional[str] = None


class ClarifyAnswerRequest(BaseModel):
    selected_choice: Optional[str] = None
    free_text: Optional[str] = None
    responded_by: str = "user"


class ResourceRef(BaseModel):
    type: str
    id: str
    name: str = ""


class ResourceScope(BaseModel):
    environment: str = "unknown"
    resources: list[ResourceRef] = Field(default_factory=list)


class ApprovalRequest(BaseModel):
    approval_id: str
    run_id: str
    summary: str
    domain: IntentDomain
    action: str
    risk_level: RiskLevel
    resource_scope: ResourceScope
    command_preview: list[str] = Field(default_factory=list)
    plan_steps: list[str] = Field(default_factory=list)
    rollback_plan: list[str] = Field(default_factory=list)
    allowed_scopes: list[ApprovalScope] = Field(default_factory=lambda: ["once"])
    expires_at: str
    created_at: str
    created_by: str = "system"


class ApprovalResponse(BaseModel):
    approval_id: str
    status: InteractionStatus = "pending"
    decision: Optional[ApprovalDecisionOutcome] = None
    final_scope: Optional[ApprovalScope] = None
    comment: Optional[str] = None
    approved_by: Optional[str] = None
    approved_at: Optional[str] = None
    next_action: Optional[str] = None


class ApprovalCreateRequest(BaseModel):
    run_id: str
    summary: str
    domain: IntentDomain
    action: str
    risk_level: RiskLevel
    resource_scope: ResourceScope
    command_preview: list[str] = Field(default_factory=list)
    plan_steps: list[str] = Field(default_factory=list)
    rollback_plan: list[str] = Field(default_factory=list)
    allowed_scopes: list[ApprovalScope] = Field(default_factory=lambda: ["once"])
    expires_in_seconds: int = 900


class ApprovalRecord(ApprovalRequest):
    status: InteractionStatus = "pending"
    decision: Optional[ApprovalDecisionOutcome] = None
    final_scope: Optional[ApprovalScope] = None
    comment: Optional[str] = None
    approved_by: Optional[str] = None
    approved_at: Optional[str] = None


class ApprovalDecision(BaseModel):
    decision: ApprovalDecisionOutcome
    scope: Optional[ApprovalScope] = None
    comment: Optional[str] = None
    approved_by: str = "user"
    approved_at: str


class ApprovalDecisionRequest(BaseModel):
    decision: ApprovalDecisionOutcome
    scope: Optional[ApprovalScope] = None
    comment: Optional[str] = None
    approved_by: str = "user"


class InteractionEnvelope(BaseModel):
    kind: InteractionKind
    clarify: Optional[ClarifyRecord] = None
    approval: Optional[ApprovalRecord] = None
    next_action: Optional[str] = None
